scala/src/com/google/idea/blaze/scala/libraries/BlazeScalaAttachSourceProvider.java

```diff
@@ -0,0 +1,178 @@
+      if (AttachedSourceJarManager.getInstance(project).hasSourceJarAttached(libraryKey)) {
+        continue;
+      }
+      BlazeJarLibrary blazeLibrary =
+          findLibraryFromIntellijLibrary(project, blazeProjectData, library);
+      if (blazeLibrary == null) {
+        continue;
+      }
+      LibraryArtifact libraryArtifact = blazeLibrary.libraryArtifact;
+      if (libraryArtifact.getSourceJars().isEmpty()) {
+        continue;
+      }
+      librariesToAttachSourceTo.add(blazeLibrary);
+    }
+
+    if (librariesToAttachSourceTo.isEmpty()) {
+      return ImmutableList.of();
+    }
+
+    // Hack: When sources are requested and we have them, we attach them automatically in the
+    // background.
+    if (attachAutomatically.getValue()) {
+      TransactionGuard.getInstance()
+          .submitTransactionLater(
+              project,
+              () -> {
+                attachSources(project, blazeProjectData, librariesToAttachSourceTo);
+              });
+      return ImmutableList.of();
+    }
+
+    return ImmutableList.of(
+        new AttachSourcesActionAdapter() {
+          @Override
+          public String getName() {
+            return "Attach Blaze Source Jars";
+          }
+
+          @Override
+          public String getBusyText() {
+            return "Attaching source jars...";
+          }
+
+          @Override
+          public ActionCallback adapterPerform(List<? extends LibraryOrderEntry> orderEntriesContainingFile) {
+            ActionCallback callback =
+                new ActionCallback().doWhenDone(() -> navigateToSource(psiFile));
+            Transactions.submitTransaction(
+                project,
+                () -> {
+                  attachSources(project, blazeProjectData, librariesToAttachSourceTo);
+                  callback.setDone();
+                });
+            return callback;
+          }
+        });
+  }
+
+  private static void navigateToSource(PsiFile psiFile) {
+    ApplicationManager.getApplication()
+        .invokeLater(
+            () -> {
+              PsiFile psi = refreshPsiFile(psiFile);
+              if (psi != null && psi.canNavigate()) {
+                psi.navigate(false);
+              }
+            });
+  }
+
+  /**
+   * The previous {@link PsiFile} can be invalidated when source jars are attached.
+   */
+  @Nullable
+  private static PsiFile refreshPsiFile(PsiFile psiFile) {
+    return PsiManager.getInstance(psiFile.getProject())
+        .findFile(psiFile.getViewProvider().getVirtualFile());
+  }
+
+  private static void attachSources(
+      Project project,
+      BlazeProjectData blazeProjectData,
+      Collection<BlazeLibrary> librariesToAttachSourceTo) {
+    ApplicationManager.getApplication()
+        .runWriteAction(
+            () -> {
+              IdeModifiableModelsProvider modelsProvider =
+                  BaseSdkCompat.createModifiableModelsProvider(project);
```

Comment:
This doesn't work on latest master, this method has been removed. I think an equivalent is 

```
new IdeModifiableModelsProviderImpl(project);
```